website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,196 @@
+    strs map(VARCHAR, VARCHAR)
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+This table has a map of 12 integers in `ints` and 5 strings in `strs`. The key is the column ordinal number in the original `lineitem` table as an integer in `ints` and as a string in `strs`.
+
+Like before, the tables are in ORC V2 and are compressed with Snappy. We show the Aria and baseline times as wall time seconds / CPU seconds, labeled with Aria: and Baseline: respectively. The queries were run on a desktop machine with two sockets and four hyperthreaded Skylake cores per socket clocked at 3.5GHz.
+
+First we compare performance of top level columns to performance of columns embedded in a non-null struct:
+```sql
+SELECT COUNT(*)
+FROM lineitem
+WHERE partkey BETWEEN 1000000 AND 2000000 AND suppkey BETWEEN 100000 AND 200000 AND extendedprice > 0;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 35                 | 1.0                     |
+| Baseline | 7                   | 80                 | 2.28                    |
+
+```sql
+SELECT COUNT(*)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 1000000 AND 2000000 AND shipment.suppkey BETWEEN 100000 AND 200000 AND shipment.extendedprice > 0 ;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 35                 | 1.0                     |
+| Baseline | 16                  | 227                | 6.5                     |
+
+We notice that for Aria it makes no difference whether the filtered columns are top level or in a non-null struct. We also note that none of the columns are materialize by Aria, since these are only filtered on, but they are materialized by baseline Presto.
+
+```sql
+SELECT COUNT(*), SUM(shipment.extendedprice)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 1000000 AND 2000000 AND shipment.suppkey BETWEEN 100000 AND 200000;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 34                 | 1.0                     |
+| Baseline | 18                  | 253                | 7.44                    |
+
+Now, instead of having a filter that is always true, we retrieve the value of `extendedprice` and materialize a struct. This takes only marginally longer because the struct only has the `extendedprice` column filled in. The struct needs to be materialized because it is an input to the SUM aggregate. We could of course, since we only access fields of the struct and not the whole struct, elide materializing the struct and only materialize the component columns. But the gain of this last optimization would not be very large.
```

Comment:
This query is doing more work than previous one, but CPU time for Aria is less. Is this due to the error of measurement. If so, perhaps, we should include that in the results. Otherwise, results look fishy.

Reply:
It's doing 1% more? That's the reason the result is within the margin of error.

Reply:
Added a commit explaining that Aria only has to materialize the struct for 1% of the values and the impact is within the margin of error for the query.